x-pack/plugins/alerting/server/rules_client/rules_client.ts

```diff
@@ -1695,6 +1715,183 @@ export class RulesClient {
+        alertingAuthorizationFilterOpts
+      );
+      return authorizationTuple.filter;
+    } catch (error) {
+      this.auditLogger?.log(
+        ruleAuditEvent({
+          action: RuleAuditAction.DELETE,
+          error,
+        })
+      );
+      throw error;
+    }
+  };
+
+  public bulkDeleteRules = async (options: BulkDeleteOptions) => {
+    const filter = (options as BulkDeleteOptionsFilter).filter;
+    const ids = (options as BulkDeleteOptionsIds).ids;
+
+    if (ids && filter) {
+      throw Boom.badRequest(
+        "Both 'filter' and 'ids' are supplied. Define either 'ids' or 'filter' properties in method arguments"
+      );
+    }
+
+    const kueryNodeFilter = ids ? convertRuleIdsToKueryNode(ids) : buildKueryNodeFilter(filter);
+    const authorizationFilter = await this.getAuthorizationFilter();
+
+    const kueryNodeFilterWithAuth =
+      authorizationFilter && kueryNodeFilter
+        ? nodeBuilder.and([kueryNodeFilter, authorizationFilter as KueryNode])
+        : kueryNodeFilter;
+
+    const { aggregations, total } = await this.unsecuredSavedObjectsClient.find<
+      RawRule,
+      RuleBulkOperationAggregation
+    >({
+      filter: kueryNodeFilterWithAuth,
+      page: 1,
+      perPage: 0,
+      type: 'alert',
+      aggs: {
+        alertTypeId: {
+          multi_terms: {
+            terms: [
+              { field: 'alert.attributes.alertTypeId' },
+              { field: 'alert.attributes.consumer' },
+            ],
+          },
+        },
+      },
+    });
+
+    if (total > MAX_RULES_NUMBER_FOR_BULK_OPERATION) {
+      throw Boom.badRequest(
+        `More than ${MAX_RULES_NUMBER_FOR_BULK_OPERATION} rules matched for bulk delete`
+      );
+    }
+
+    const buckets = aggregations?.alertTypeId.buckets;
+
+    if (buckets === undefined || buckets?.length === 0) {
+      throw Boom.badRequest('No rules found for bulk delete');
+    }
+
+    await pMap(
+      buckets,
+      async ({ key: [ruleType, consumer] }) => {
+        this.ruleTypeRegistry.ensureRuleTypeEnabled(ruleType);
+        try {
+          await this.authorization.ensureAuthorized({
+            ruleTypeId: ruleType,
+            consumer,
+            operation: WriteOperations.BulkDelete,
+            entity: AlertingAuthorizationEntity.Rule,
+          });
+        } catch (error) {
+          this.auditLogger?.log(
+            ruleAuditEvent({
+              action: RuleAuditAction.DELETE,
+              error,
+            })
+          );
+          throw error;
+        }
+      },
+      { concurrency: RULE_TYPE_CHECKS_CONCURRENCY }
+    );
+
+    const { apiKeysToInvalidate, errors, taskIdsToDelete } = await retryIfBulkDeleteConflicts(
+      this.logger,
+      (filterKueryNode: KueryNode | null) => this.bulkDeleteWithOCC({ filter: filterKueryNode }),
+      kueryNodeFilterWithAuth
+    );
+
+    if (taskIdsToDelete.length > 0) {
+      this.taskManager.bulkRemoveIfExist(taskIdsToDelete);
```

Comment:
I thought, we do not need to wait for return here and it is not important to wait for execution to be completed before a next step , so I skipped await.

Reply:
If we do not use `await` and `taskManager.bulkRemoveIfExists` throws an error, it will cause an Unhandled promise rejection and crash Kibana, requiring a restart. Why do we not feel it is important to let users know that something went wrong deleting their task documents?

Reply:
This would actually be good to add a unit test for...testing what happens when `taskManager.bulkRemoveIfExists` throws an error.